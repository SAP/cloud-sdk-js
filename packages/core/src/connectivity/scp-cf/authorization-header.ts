import {
  unixEOL,
  isNullish,
  createLogger,
  pickIgnoreCase,
  toSanitizedObject,
  encodeBase64
} from '@sap-cloud-sdk/util';
import type { ODataRequest, ODataRequestConfig } from '../../odata-common';
import {
  AuthenticationType,
  Destination,
  DestinationAuthToken,
  sanitizeDestination
} from './destination';

const logger = createLogger({
  package: 'core',
  messageContext: 'authorization-header'
});

/**
 * @deprecated Since v1.20.0. Use [[buildAuthorizationHeaders]] instead.
 * Adds authorization headers for a given ODataRequest to existing headers.
 *
 * @param request - an ODataRequest.
 * @param headers - The headers that should be added to.
 * @returns The provided headers with the new authorization headers.
 */
export async function addAuthorizationHeader<
  RequestT extends ODataRequestConfig
>(
  request: ODataRequest<RequestT>,
  headers: Record<string, string>
): Promise<Record<string, string>> {
  const destination = request.destination;
  if (!destination) {
    return headers;
  }
  const authHeaders = await getAuthHeaders(
    destination,
    request.config.customHeaders
  );
  return {
    ...headers,
    ...authHeaders
  };
}

function needsAuthHeaders(destination: Destination): boolean {
  if (!destination.authentication) {
    return false;
  }
  const authTypesWithAuthorizationHeader: AuthenticationType[] = [
    'BasicAuthentication',
    'OAuth2ClientCredentials',
    'OAuth2SAMLBearerAssertion',
    'OAuth2UserTokenExchange',
    'OAuth2JWTBearer',
    'PrincipalPropagation'
  ];
  return authTypesWithAuthorizationHeader.includes(destination.authentication);
}

function getCustomAuthHeaders(
  destination: Destination,
  customHeaders?: Record<string, any>
) {
  if (destination.authentication === 'PrincipalPropagation') {
    return pickIgnoreCase(customHeaders, 'SAP-Connectivity-Authentication');
  }
  return pickIgnoreCase(customHeaders, 'authorization');
}

export async function getAuthHeaders(
  destination: Destination,
  customHeaders?: Record<string, any>
): Promise<Record<string, string>> {
  const customAuthHeaders = getCustomAuthHeaders(destination, customHeaders);

  if (Object.keys(customAuthHeaders).length && needsAuthHeaders(destination)) {
    logger.warn(
      'Found custom authorization headers. The given destination also provides authorization headers. This might be unintended. The custom headers from the request config will be used.'
    );
  }

  const additionalDestinationAuthHeaders = getCustomAuthHeaders(
    destination,
    destination.headers
  );

  if (Object.keys(customAuthHeaders).length) {
    return customAuthHeaders;
  }
  if (Object.keys(additionalDestinationAuthHeaders).length) {
    return additionalDestinationAuthHeaders;
  }
  return buildAuthorizationHeaders(destination);
}

/**
 * @deprecated Since v1.20.0. Use [[buildAuthorizationHeaders]] instead.
 * Adds authorization headers for a given destination to existing headers.
 *
 * @param destination - A destination.
 * @param headers - The headers that should be added to.
 * @returns The provided headers with the new authorization headers.
 */
export function buildAndAddAuthorizationHeader(destination: Destination) {
  return async function (
    headers: Record<string, any>
  ): Promise<Record<string, string>> {
    return {
      ...headers,
      ...(await buildAuthorizationHeaders(destination))
    };
  };
}

function toAuthorizationHeader(authorization: string): Record<string, string> {
  return toSanitizedObject('authorization', authorization);
}

function headerFromTokens(
  authenticationType: AuthenticationType,
  authTokens?: DestinationAuthToken[] | null
): Record<string, string> {
  if (!authTokens || !authTokens.length) {
    throw Error(
      `\`AuthenticationType\` is "${authenticationType}", but no auth tokens could be fetched from the destination service.`
    );
  }
  const usableTokens = authTokens.filter(
    (token: DestinationAuthToken) => !token.error
  );

  if (!usableTokens.length) {
    throw Error(
      [
        'The destination tried to provide authorization tokens but failed in all cases. This is most likely due to misconfiguration.',
        'Original error messages:',
        ...authTokens.map(token => token.error)
      ].join(unixEOL)
    );
  }
  const authToken = usableTokens[0];
  return toAuthorizationHeader(authToken.http_header.value);
}

function headerFromBasicAuthDestination(
  destination: Destination
): Record<string, string> {
  if (isNullish(destination.username) || isNullish(destination.password)) {
    throw Error(
      'AuthenticationType is "BasicAuthentication", but "username" and / or "password" are missing!'
    );
  }

  return toAuthorizationHeader(
    basicHeader(destination.username, destination.password)
  );
}

export function basicHeader(username: string, password: string): string {
  return 'Basic ' + encodeBase64(`${username}:${password}`);
}

function headerForPrincipalPropagation(
  destination: Destination
): Record<string, any> {
  const principalPropagationHeader =
    destination?.proxyConfiguration?.headers?.[
      'SAP-Connectivity-Authentication'
    ];
  if (!principalPropagationHeader) {
    throw Error(
      'Principal propagation was selected in destination, but no SAP-Connectivity-Authentication bearer header was added by connectivity service.'
    );
  }
  return {
    'SAP-Connectivity-Authentication': principalPropagationHeader
  };
}

function headerForProxy(destination: Destination): Record<string, any> {
  const proxyAuthHeader =
    destination?.proxyConfiguration?.headers?.['Proxy-Authorization'];

  return proxyAuthHeader ? { 'Proxy-Authorization': proxyAuthHeader } : {};
}

// TODO the proxy header are for OnPrem auth and are now handled correctly and should be removed here
// However this would be a breaking change, since we recommended to use 'NoAuthentication' to achieve principal propagation as a workaround.
// Remove this in v2
function legacyNoAuthOnPremiseProxy(
  destination: Destination
): Record<string, any> {
  logger.warn(
    `You are using 'NoAuthentication' in destination: ${destination.name} which is an OnPremise destination. This is a deprecated configuration, most likely you wanted to set-up 'PrincipalPropagation' so please change the destination property to the desired authentication scheme.`
  );

  let principalPropagationHeader;
  try {
    principalPropagationHeader = headerForPrincipalPropagation(destination);
  } catch (e) {
    logger.warn('No principal propagation header found.');
  }

  return {
    ...headerForProxy(destination),
    ...principalPropagationHeader
  };
}

function getProxyRelatedAuthHeaders(
  destination: Destination
): Record<string, any> {
  if (
    destination.proxyType === 'OnPremise' &&
    destination.authentication === 'NoAuthentication'
  ) {
    return legacyNoAuthOnPremiseProxy(destination);
  }

  // The connectivity service will raise an exception if it can not obtain the 'Proxy-Authorization' and the destination lookup will fail early
  return headerForProxy(destination);
}

async function getAuthenticationRelatedHeaders(
  destination: Destination
): Promise<Record<string, string>> {
  logger.debug(
    `Getting authentication related headers for authentication type: ${destination.authentication}`
  );

  switch (destination.authentication) {
    case null:
    case undefined:
      logger.warn(
        'No authentication type is specified on the destination! Assuming "NoAuthentication".'
      );
      return {};
    case 'NoAuthentication':
    case 'ClientCertificateAuthentication':
      return {};
    case 'OAuth2SAMLBearerAssertion':
    case 'OAuth2UserTokenExchange':
    case 'OAuth2JWTBearer':
    case 'OAuth2ClientCredentials':
      return headerFromTokens(
        destination.authentication,
        destination.authTokens
      );
    case 'BasicAuthentication':
      return headerFromBasicAuthDestination(destination);
    case 'PrincipalPropagation':
      return headerForPrincipalPropagation(destination);
    default:
      throw Error(
        `The destination used "${destination.authentication}" as authentication type which is not supported by the SAP Cloud SDK.`
      );
  }
}

export async function buildAuthorizationHeaders(
  destination: Destination
): Promise<Record<string, string>> {
  const sanitizedDestination = sanitizeDestination(destination);
  return {
    ...(await getAuthenticationRelatedHeaders(sanitizedDestination)),
    ...getProxyRelatedAuthHeaders(sanitizedDestination)
  };
}
