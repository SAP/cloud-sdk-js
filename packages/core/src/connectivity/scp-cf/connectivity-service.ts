import { createLogger, ErrorWithCause } from '@sap-cloud-sdk/util';
import { Protocol } from './protocol';
import { ProxyConfiguration } from './connectivity-service-types';
import {
  AuthenticationType,
  Destination
} from './destination/destination-service-types';
import { EnvironmentAccessor } from './environment-accessor';
import { Service } from './environment-accessor-types';
import { serviceToken } from './token-accessor';
import { decodeJwt, isUserToken, JwtPair } from './jwt';

const logger = createLogger({
  package: 'core',
  messageContext: 'connectivity-service'
});

/**
 * Given a destination and a JWT (required for subscriber destinations), this function will add a proxy configuration to a destination.
 * See also [[ProxyConfiguration]].
 *
 * This function will reject if no connectivity service is bound, no XSUAA service with plan application is bound or the client credentials grant with the XSUAA service fails.
 *
 * @Deprecated Since v1.16.0. Use [[addProxyConfigurationOnPrem]] instead.
 * @param destination - The destination to which the proxy configuration should be added.
 * @param jwt - The JWT of the current user.
 * @returns A promise resolving to the destination with the added proxy configuration.
 */
export function addProxyConfiguration(
  destination: Destination,
  jwt?: string
): Promise<Destination> {
  return Promise.resolve()
    .then(() => proxyHostAndPort())
    .then(hostAndPort =>
      addHeaders(hostAndPort, destination.authentication, jwt)
    )
    .then(proxyConfiguration => ({
      ...destination,
      proxyConfiguration
    }));
}

// TODO: remove string argument in v2.0
export function addProxyConfigurationOnPrem(
  destination: Destination,
  jwt: string | JwtPair | undefined
): Promise<Destination> {
  const jwtPair =
    typeof jwt === 'string' ? { encoded: jwt, decoded: decodeJwt(jwt) } : jwt;
  if (
    destination.authentication === 'PrincipalPropagation' &&
    !isUserToken(jwtPair)
  ) {
    throw new Error('For principal propagation a user JWT is needed.');
  }

  return addProxyConfiguration(destination, jwtPair?.encoded);
}

interface HostAndPort {
  host: string;
  port: number;
  protocol: Protocol.HTTP;
}

export function proxyHostAndPort(): HostAndPort {
  const service = readConnectivityServiceBinding();
  return {
    host: service.credentials.onpremise_proxy_host,
    port:
      service.credentials.onpremise_proxy_http_port ||
      service.credentials.onpremise_proxy_port,
    protocol: Protocol.HTTP
  };
}

function readConnectivityServiceBinding(): Service {
  const serviceBindings = EnvironmentAccessor.getServiceList('connectivity');

  if (!serviceBindings.length) {
    throw new Error(
      'No binding to a connectivity service found! Please make sure to bind an instance of the connectivity service to your app if you want to connect to on-premise destinations.'
    );
  }

  return serviceBindings[0];
}

function addHeaders(
  hostAndPort: HostAndPort,
  authenticationType: AuthenticationType | undefined,
  jwt?: string
): Promise<ProxyConfiguration> {
  const connServiceBinding = readConnectivityServiceBinding();

  return Promise.resolve()
    .then(() => proxyAuthorizationHeader(connServiceBinding, jwt))
    .then(proxyAuthHeader => ({
      ...proxyAuthHeader,
      ...sapConnectivityAuthenticationHeader(authenticationType, jwt)
    }))
    .then(
      headers =>
        ({
          ...hostAndPort,
          headers
        } as ProxyConfiguration)
    );
}

function proxyAuthorizationHeader(
  connectivityServiceBinding: Service,
  userJwt?
): Promise<Record<string, string>> {
  return serviceToken(connectivityServiceBinding, { userJwt })
    .then(token => ({
      'Proxy-Authorization': `Bearer ${token}`
    }))
    .catch(error => {
      throw new ErrorWithCause(
        'Failed to add proxy authorization header - client credentials grant failed!',
        error
      );
    });
}

function sapConnectivityAuthenticationHeader(
  authenticationType: AuthenticationType | undefined,
  jwt?: string
): Record<string, string> {
  if (authenticationType === 'PrincipalPropagation') {
    if (jwt) {
      return {
        'SAP-Connectivity-Authentication': `Bearer ${jwt}`
      };
    }
    throw new Error(
      `Unable to create "SAP-Connectivity-Authentication" header: no JWT found on the current request.
     Connecting to on-premise systems via principle propagation is not possible.`
    );
  }
  if (authenticationType === 'BasicAuthentication') {
    logger.warn(
      'You are connecting to an On-Premise system using basic authentication. For productive usage Principal propagation is recommended.'
    );
  }

  return {};
}
