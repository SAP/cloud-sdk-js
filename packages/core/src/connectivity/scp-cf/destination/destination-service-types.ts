import { CachingOptions, IsolationStrategy } from '../cache';
import { ProxyConfiguration } from '../connectivity-service-types';
import { ResilienceOptions } from '../resilience-options';

/**
 * A resolved destination containing information needed to execute requests, such as the system URL.
 *
 * You can create a destination as a local object when supplying all necessary information, or it could be retrieved from the destination service on SAP Cloud Platform (via [[DestinationNameAndJwt]]).
 * When creating a local object representing a destination, you need to supply at least the [[url]] and, if required by the target system, valid credentials with [[username]] and [[password]].
 */
export interface Destination {
  /**
   * Name of the destination retrieved from SAP Cloud Platform.
   */
  name?: string | null;

  /**
   * Base URL for calls to this destination.
   * The URL has to define the protocol, like `http://` or `https://`, and a host.
   * The path for requests against this destination will be appended to the path defined in the URL as a new path segment.
   */
  url: string;

  /**
   * Type of authentication to use.
   *
   * Defaults to `NoAuthentication`, unless [[username]] and [[password]] are provided, in which case the default is `BasicAuthentication`.
   */
  authentication?: AuthenticationType;

  /**
   * Proxy type to specify whether the target resides on-premise (not used).
   */
  proxyType?: DestinationProxyType;

  /**
   * @deprecated Since v1.0.1.
   * Origin of the destination in a multi-tenant setup on SAP Cloud Platform (either from provider or subscriber account), optional.
   */
  origin?: DestinationOrigin;

  /**
   * Client to target in an SAP system, will be added as HTTP header `sap-client` if set.
   */
  sapClient?: string | undefined | null;

  /**
   * Username to use for basic authentication, optional if other means of authentication shall be used.
   */
  username?: string | null;

  /**
   * Password to use for basic authentication, optional if other means of authentication shall be used.
   */
  password?: string | null;

  /**
   * Authentication tokens returned from destination service on SAP Cloud Platform.
   */
  authTokens?: DestinationAuthToken[] | null;

  /**
   * Flag indicating whether all certificates should be accepted when communicating with the destination. Should not be "true" in production.
   */
  isTrustingAllCertificates?: boolean;

  /**
   * ProxyConfiguration for on-premise connectivity and http(s) web proxies. Is present if proxyType of the destination equals "OnPremise" or environment variables [http_proxy] or [https_proxy] are set See [[ProxyConfiguration]].
   */
  proxyConfiguration?: ProxyConfiguration;

  /**
   * Client Id used to retrieve access token for "OAuth2ClientCredentials", "OAuth2UserTokenExchange" and "OAuth2JWTBearer" authentication.
   */
  clientId?: string;

  /**
   * Client Secret used to retrieve access token for "OAuth2ClientCredentials", "OAuth2UserTokenExchange" and "OAuth2JWTBearer" authentication.
   */
  clientSecret?: string;

  /**
   * URL to retrieve access token for "OAuth2ClientCredentials", "OAuth2UserTokenExchange" and "OAuth2JWTBearer" authentication.
   */
  tokenServiceUrl?: string;

  /**
   * User for basic authentication to OAuth server (if required).
   */
  tokenServiceUser?: string;

  /**
   * Password for tokenServiceUser (if required).
   */
  tokenServicePassword?: string;

  /**
   * The type of the destination, defaults to 'HTTP'. The SAP Cloud SDK only understands destinations of type 'HTTP'.
   */
  type?: 'HTTP' | 'LDAP' | 'MAIL' | 'RFC';

  /**
   * Further properties of the destination as defined in destination service on SAP Cloud Platform, possibly empty.
   */
  originalProperties?: { [key: string]: any };

  /**
   * Flag indicating whether the destination is for test purpose. Should be "undefined" or "false" for non-mocked destinations.
   */
  isTestDestination?: boolean;

  /**
   * Location ID of the Cloud Connector to be used for connection to an On-Premise system. Optional. Corresponds to property "CloudConnectorLocationId" in the additional properties of a destination.
   */
  cloudConnectorLocationId?: string;

  /**
   * Array of certificates used for authentication type ClientCertificateAuthentication.
   */
  certificates?: DestinationCertificate[];

  /**
   * Name of the key store/certificate to be used for ClientCertificateAuthentication.
   */
  keyStoreName?: string;

  /**
   * Password of the key store/certificate to be used for ClientCertificateAuthentication.
   */
  keyStorePassword?: string;

  /**
   * System user to be used for OAuth2SAMLBearerAssertion authentication type.
   */
  systemUser?: string;

  /**
   * Additional headers to be used for calls against the destination, originally defined by `URL.headers.<header-name>`.
   * The keys of this object denote the names of the headers and the values their values.
   */
  headers?: Record<string, any>;

  /**
   * Additional query parameters to be used for calls against the destination, originally defined by `URL.queries.<query-parameter-name>`.
   * The keys of this object denote the names of the query parameters and the values their values.
   */
  queryParameters?: Record<string, any>;

  /**
   * If set to true the auth token provided to the request execution is forwarded to the destination target.
   */
  forwardAuthToken?: boolean;
}

/**
 * Represents authentication token returned from destination service.
 */
export interface DestinationAuthToken {
  type: string;
  value: string;
  expiresIn: string;
  error: string | null;
  http_header: {
    key: string;
    value: string;
  };
}

export type DestinationProxyType = 'OnPremise' | 'Internet' | null;

/**
 * Represents a certificate attached to a destination.
 */
export interface DestinationCertificate {
  /**
   * Name of the certificate file.
   */
  name: string;

  /**
   * Content of the certificate as base64 encoded binary.
   */
  content: string;

  /**
   * Type of the certificate.
   */
  type: string;
}

/**
 * @deprecated Since v1.0.1.
 *
 * Represents the origin of a destination in a multi-tenant setup on SAP Cloud Platform.
 *
 * In a multi-tenant application on SAP Cloud Platform, destinations can be defined both on provider account level ("PaaS tenant") as well as on the level of each subscriber account ("SaaS tenant").
 */
export enum DestinationOrigin {
  Subscriber = 'subscriber',
  Provider = 'provider'
}

/**
 * Declaration of a destination to be retrieved from an environment variable or from the destination service on SAP Cloud Platform.
 *
 * Use an object of this interface to specify which destination shall be used when executing a request.
 * The destination will be retrieved via its [[DestinationNameAndJwt.destinationName]] according to the following algorithm:
 * 1. If a destination of this [[DestinationNameAndJwt.destinationName]] is defined in the environment variable `destinations` (if available), it will be converted into a [[Destination]] and used for the request.
 * 2. Otherwise, the destination service on SAP Cloud Platform is queried for a destination with the given [[DestinationNameAndJwt.destinationName]], using the access token provided as value of property [[jwt]].
 */
export interface DestinationNameAndJwt {
  /**
   * Name of the destination to retrieve, mandatory.
   */
  destinationName: string;

  /**
   * An access token for the XSUAA service on SAP Cloud Platform, provided as a JSON Web Token, only mandatory when destination shall be retrieved from destination service on SAP Cloud Platform.
   */
  jwt?: string;
}

/**
 * @deprecated Since version 1.16.0. Use [[CachingOptions]] instead.
 */
export interface DestinationCachingOptions {
  /**
   * A boolean value that indicates whether to read destinations from cache.
   */
  useCache?: boolean;

  /**
   * The isolation strategy used for caching destinations. For the available options, see [[IsolationStrategy]].
   * By default, IsolationStrategy.Tenant is set.
   */
  isolationStrategy?: IsolationStrategy;
}

/**
 * Options to use while fetching destinations. Encompasses both [[DestinationCachingOptions]] and [[ResilienceOptions]] interfaces.
 */
export type DestinationRetrievalOptions = (
  | DestinationCachingOptions
  | CachingOptions
) &
  ResilienceOptions;

export function isDestinationNameAndJwt(
  destination: any
): destination is DestinationNameAndJwt {
  return destination.destinationName !== undefined;
}

export function isDestination(destination: any): destination is Destination {
  return destination.url !== undefined;
}

export type AuthenticationType =
  | 'PrincipalPropagation'
  | 'NoAuthentication'
  | 'BasicAuthentication'
  | 'OAuth2SAMLBearerAssertion'
  | 'OAuth2ClientCredentials'
  | 'OAuth2UserTokenExchange'
  | 'ClientCertificateAuthentication'
  | 'OAuth2JWTBearer'
  | 'OAuth2Password';

/**
 * The destinations endpoint distinguished between destinations maintained on service level (instance) and account level (subaccount).
 * This enum is used as a switch in the [[fetchInstanceDestinations]], [[fetchSubaccountDestinations]]  and [[destinationServiceCache]]
 */
export enum DestinationType {
  Instance = 'instance',
  Subaccount = 'subaccount'
}
