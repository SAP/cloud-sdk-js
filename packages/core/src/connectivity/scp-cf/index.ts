export * from './jwt';
export * from './cache';
export * from './client-credentials-token';
export * from './client-credentials-token-cache';
export * from './connectivity-service';
export * from './connectivity-service-types';
export * from './destination';
export * from './environment-accessor';
export * from './environment-accessor-types';
export * from './resilience-options';
export * from './token-accessor';
export * from './xsuaa-service';
export * from './xsuaa-service-types';
export * from './protocol';
export * from './authorization-header';
export * from './get-protocol';
export * from './header-builder-for-destination';
export {
  mappingTenantFields,
  Tenant,
  tenantFromJwt,
  RegisteredJWTClaimsTenant,
  isIdenticalTenant
} from './tenant';
export {
  RegisteredJWTClaimsUser,
  Scope,
  mappingUserFields,
  userFromJwt,
  customAttributes
} from './user';
