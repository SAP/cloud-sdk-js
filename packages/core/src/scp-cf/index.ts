export * from './cache';
export * from './client-credentials-token';
export * from './client-credentials-token-cache';
export * from './connectivity-service';
export * from './connectivity-service-types';
export * from './destination';
export * from './destination-accessor';
export * from './destination-accessor-types';
export * from './destination-cache';
export * from './destination-selection-strategies';
export * from './destination-service';
export * from './destination-service-types';
export * from './env-destination-accessor';
export * from './environment-accessor';
export * from './environment-accessor-types';
export * from './resilience-options';
export * from './token-accessor';
export * from './vcap-service-destination';
export * from './xsuaa-service';
export * from './xsuaa-service-types';
