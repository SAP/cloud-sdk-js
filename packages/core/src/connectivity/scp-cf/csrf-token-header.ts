import {
  createLogger,
  ErrorWithCause,
  pickIgnoreCase,
  pickNonNullish,
  pickValueIgnoreCase
} from '@sap-cloud-sdk/util';
import { AxiosError } from 'axios';
import { HttpRequestConfig, executeHttpRequest } from '../../http-client';
import { Destination, DestinationNameAndJwt } from '../scp-cf';
import { removeTrailingSlashes } from '../../odata-common/remove-slashes';

const logger = createLogger({
  package: 'core',
  messageContext: 'csrf-token-header'
});

/**
 * Get CSRF token and cookies for a destination and request configuration. The CSRF token and cookies will be retrieved based on the url of the destination and the custom configuration given by the `requestConfig`.
 * If there is a relative url in the `requestConfig` it will be appended to the destination's url, an absolute url overwrites the destination related url.
 * @param destination The destination to get the headers from
 * @param requestConfig An http request configuration containing additional information about the request, like url or headers
 * @returns A promise to an object containing the CSRF related headers
 */
export async function buildCsrfHeaders<T extends HttpRequestConfig>(
  destination: Destination | DestinationNameAndJwt,
  requestConfig: Partial<T>
): Promise<Record<string, any>> {
  const csrfHeaders = await makeCsrfRequest(destination, requestConfig);
  validateCsrfTokenResponse(csrfHeaders);
  return pickNonNullish({
    ...pickIgnoreCase(csrfHeaders, 'x-csrf-token'),
    cookie: buildCookieHeaderValue(
      pickValueIgnoreCase(csrfHeaders, 'set-cookie')
    )
  });
}

function makeCsrfRequest<T extends HttpRequestConfig>(
  destination: Destination | DestinationNameAndJwt,
  requestConfig: Partial<T>
): Promise<Record<string, any>> {
  const fetchHeader = !pickValueIgnoreCase(
    requestConfig.headers,
    'x-csrf-token'
  ) && { 'x-csrf-token': 'Fetch' };
  const axiosConfig: HttpRequestConfig = {
    method: 'get',
    ...requestConfig,
    headers: {
      ...fetchHeader,
      ...requestConfig.headers
    },
    url: requestConfig.url
  };

  // The S/4 does a redirect if the CSRF token is fetched in case the '/' is not in the URL.
  // TODO: remove once https://github.com/axios/axios/issues/3369 is really fixed. Issue is closed but problem stays.
  return executeHttpRequest(destination, appendSlash(axiosConfig))
    .then(response => response.headers)
    .catch(e1 => {
      logger.error(
        new ErrorWithCause(
          'Initial try to fetch CSRF token failed - retry without slash at ',
          e1
        )
      );
      return executeHttpRequest(destination, removeSlash(axiosConfig))
        .then(response => response.headers)
        .catch(e2 => {
          throw new ErrorWithCause(
            'Also second try to fetch SRF token failed - No CSRF token fetched.',
            e2
          );
        });
    });
}

function appendSlash(requestConfig: HttpRequestConfig): HttpRequestConfig {
  if (!requestConfig.url!.endsWith('/')) {
    requestConfig.url = `${requestConfig.url}/`;
  }
  return requestConfig;
}

function removeSlash(requestConfig: HttpRequestConfig): HttpRequestConfig {
  if (requestConfig.url!.endsWith('/')) {
    requestConfig.url = removeTrailingSlashes(requestConfig.url!);
  }
  return requestConfig;
}

function axiosWorkaround<T extends HttpRequestConfig>(
  error: AxiosError,
  axiosConfig: Partial<T>,
  destination: Destination | DestinationNameAndJwt
) {
  if (error.request?._isRedirect && error.request?._options?.path) {
    logger.warn(
      'Csrf fetch was redirected and failed. This might be a bug in the underlying request library (https://github.com/axios/axios/issues/3369).\nRetrying with full configuration.'
    );
    return makeCsrfRequest(destination, {
      ...axiosConfig,
      url: error.request._options.path
    });
  }
}

function validateCsrfTokenResponse(responseHeaders: Record<string, any>) {
  if (!responseHeaders['x-csrf-token']) {
    logger.warn(
      'Destination did not return a CSRF token. This may cause a failure when sending the OData request.'
    );
  }

  if (!responseHeaders['set-cookie']) {
    logger.warn('CSRF header response does not include cookies.');
  }

  return responseHeaders;
}

function buildCookieHeaderValue(cookies?: string[]): string | undefined {
  if (cookies && cookies.length) {
    return cookies.map((cookie: string) => cookie.split(';')[0]).join(';');
  }
}
