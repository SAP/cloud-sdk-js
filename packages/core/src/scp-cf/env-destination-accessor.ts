/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import { addProxyConfigurationInternet, ProxyStrategy, proxyStrategy } from '../util/proxy-util';
import { sanitizeDestination } from './destination';
import { Destination } from './destination-service-types';
import { createLogger } from '@sap-cloud-sdk/util';

const logger = createLogger({
  package: 'core',
  messageContext: 'env-destination-accessor'
});

/**
 * Get all destinations from the environment variable "destinations".
 * This is discouraged for productive use! Use destination-accessor/useOrFetchDestination for fetching destinations
 * from the Cloud Foundry destination service.
 *
 * @returns A list of destinations
 */
export function getDestinationsFromEnv(): Destination[] {
  return getDestinationsEnvVariable() ? (JSON.parse(getDestinationsEnvVariable()!)).map(entry => sanitizeDestination(entry)) : [];
}

/**
 * @deprecated Since v1.4.2. Use [[getDestinationsFromEnv]] instead.
 *
 * Get all destinations from the environment variable "destinations".
 * This is discouraged for productive use! Use destination-accessor/useOrFetchDestination for fetching destinations
 * from the Cloud Foundry destination service.
 *
 * @returns A list of destinations
 */
export function getDestinations(): Destination[] {
  return getDestinationsFromEnv();
}

/**
 * Get a destination from the environment variables by name. Throws an error if there are multiple destinations with the same name.
 * This is discouraged for productive use! Use destination-accessor/useOrFetchDestination for fetching destinations
 * from the Cloud Foundry destination service.
 *
 * @param name - Name of the destination
 * @returns The requested destination if existent, otherwise `null`
 */
export function getDestinationFromEnvByName(name: string): Destination | null {
  const destinations = getDestinationsFromEnv();
  if (destinations.length) {
    validateDestinations(destinations);
    const matchingDestinations = destinations.filter(dest => dest.name === name);
    if (matchingDestinations.length > 1) {
      throw new Error(`There are multiple destinations with the name '${name}'.`);
    }
    const destination = matchingDestinations[0];
    if (destination) {
      return proxyStrategy(destination) === ProxyStrategy.INTERNET_PROXY ? addProxyConfigurationInternet(destination) : destination;
    }
  }
  return null;
}

/**
 * @deprecated Since v1.4.2. Use [[getDestinationFromEnvByName]] instead.
 *
 * Get a destination from the environment variables by name. Throws an error if there are multiple destinations with the same name.
 * This is discouraged for productive use! Use destination-accessor/useOrFetchDestination for fetching destinations
 * from the Cloud Foundry destination service.
 *
 * @param name - Name of the destination
 * @returns The requested destination if existent, otherwise `null`
 */
export function getDestinationByName(name: string): Destination | null {
  return getDestinationFromEnvByName(name);
}

/* eslint-disable valid-jsdoc */

/**
 * @hidden
 */
export function getDestinationConfig(dest: string | Destination = 'ErpQueryEndpoint'): Destination | null {
  return typeof dest === 'string' ? getDestinationFromEnvByName(dest) : dest;
}

/**
 * @hidden
 */
export function getDestinationsEnvVariable(): string | undefined {
  return process.env['destinations'];
}

function validateDestinations(destinations: Destination[]) {
  destinations.forEach(destination => {
    if (typeof destination.name === 'undefined') {
      logMissingPropertyWarning(destination, 'name');
    }
    if (typeof destination.url === 'undefined') {
      logMissingPropertyWarning(destination, 'url');
    }
  })
}

function logMissingPropertyWarning(destination: Destination, property: string): void {
  logger.warn(`Destination from 'destinations' env variable is missing '${property}' property. Make sure it exists: ${JSON.stringify(destination)}`);
}
