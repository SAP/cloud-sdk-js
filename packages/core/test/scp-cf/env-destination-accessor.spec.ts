/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
import * as assert from 'assert';
import { Destination } from '../../src/scp-cf/destination-service-types';
import { getDestinationByName, getDestinations } from '../../src/scp-cf/env-destination-accessor';

const environmentDestination = {
  name: 'ErpQueryEndpoint',
  url: 'https://mys4hana.com',
  username: 'myuser',
  password: 'mypw'
};

const destinationFromEnv: Destination = {
  authTokens: [],
  authentication: 'BasicAuthentication',
  name: 'ErpQueryEndpoint',
  isTrustingAllCertificates: false,
  originalProperties: {
    name: 'ErpQueryEndpoint',
    url: 'https://mys4hana.com',
    username: 'myuser',
    password: 'mypw'
  },
  password: 'mypw',
  username: 'myuser',
  url: 'https://mys4hana.com'
};

describe('getDestinations()', () => {
  afterEach(() => {
    delete process.env.destinations;
  });

  it('should return all destinations from environment variables', () => {
    // Setup
    process.env['destinations'] = `[${JSON.stringify(environmentDestination)}]`;

    const expected: Destination[] = [destinationFromEnv];
    const actual = getDestinations();
    expected.forEach((e, index) => {
      expect(actual[index]).toMatchObject(e);
    });
  });

  it('should return an empty array if no destinations can be found', () => {
    const expected = [];
    const actual = getDestinations();
    assert.deepEqual(expected, actual, 'No empty array was returned');
  });
});

describe('getDestinationByName()', () => {
  afterEach(() => {
    delete process.env.destinations;
  });

  it('should return the destination if a destination with the respective name is present in the environment variables', () => {
    // Setup
    process.env['destinations'] = `[${JSON.stringify(environmentDestination)}]`;

    const expected: Destination = destinationFromEnv;
    const actual = getDestinationByName('ErpQueryEndpoint');
    expect(actual).toMatchObject(expected);
  });

  it('should return null when no destination can be found', () => {
    const expected = null;
    const actual = getDestinationByName('ErpQueryEndpoint');

    assert.equal(actual, expected, 'Expected null, but got something.');
  });

  it('should throw an error when multiple destinations with the same name are found', () => {
    process.env['destinations'] = `[${JSON.stringify(environmentDestination)},${JSON.stringify(environmentDestination)}]`;

    try {
      getDestinationByName('ErpQueryEndpoint');
    } catch (error) {
      assert.equal('There are multiple destinations with the name "ErpQueryEndpoint".', error.message, 'Did not get the expected error message.');
    }
  });
});
