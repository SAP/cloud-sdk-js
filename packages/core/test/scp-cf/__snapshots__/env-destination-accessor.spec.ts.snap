// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDestinationByName() should throw an error when multiple destinations with the same name are found 1`] = `"There are multiple destinations with the name 'FINAL_DESTINATION'."`;
